cirq/optimizers/three_qubit_decomposition.py

```diff
@@ -0,0 +1,272 @@
+    # next to each other
+    def num_conseq_2qbit_gates(i):
+        j = i
+        while j < len(operations) and operations[j].gate.num_qubits() == 2:
+            j += 1
+        return j - i
+
+    operations = list(circuit.all_operations())
+
+    i = 0
+    while i < len(operations):
+        num_czs = num_conseq_2qbit_gates(i)
+        if num_czs == 4:
+            operations = operations[:1]
+            break
+        elif num_czs == 3:
+            operations = operations[:i] + [operations[i + 1]] + operations[i + 3 :]
+            break
+        else:
+            i += 1
+    return operations
+
+
+def _two_qubit_multiplexor_to_circuit(
+    q0: ops.Qid,
+    q1: ops.Qid,
+    q2: ops.Qid,
+    u1: np.ndarray,
+    u2: np.ndarray,
+    shift_left: bool = True,
+    diagonal: np.ndarray = np.eye(4),
+    atol: float = 1e-8,
+):
+    r"""Converts a two qubit double multiplexor to circuit.
+    Input: u1 ⊕ u2, with select qubit a (i.e. a = |0> => u1(b,c),
+    a = |1> => u2(b,c).
+
+    We want this:
+        $$
+        u1 ⊕ u2 = v ⊕ v @ D ⊕ D^{\dagger} @ W ⊕ W
+        $$
+    We can get it via:
+        $$
+        u1 = v @ D @ W       (1)
+        u2 = v @ D^{\dagger} @ W (2)
+        $$
+
+    We can derive
+        $$
+        u1u2^{adj}= v @ D^2 @ v^{\dagger}, (3)
+        $$
+
+    i.e the eigendecomposition of $u1u2^{\dagger}$ will give us D and v.
+    W is easy to derive from (2).
+
+    This function, after calculating v, D and W, also returns the circuit that
+    implements these unitaries: v, W on qubits b, c and the middle diagonal
+    multiplexer on a,b,c qubits.
+
+    The resulting circuit will have only known two-qubit and one-qubit gates,
+    namely CZ, CNOT and rx, ry, PhasedXPow gates.
+
+    Args:
+        q0: first qubit
+        q1: second qubit
+        q2: third qubit
+        u1: two-qubit operation on b,c for a = |0>
+        u2: two-qubit operation on b,c for a = |1>
+        shift_left: return the extracted diagonal or not
+        diagonal: an incoming diagonal to be merged with
+    Returns:
+        The circuit implementing the two qubit multiplexor consisting only of
+        known two-qubit and single qubit gates
+    """
+    u1u2 = u1 @ u2.conj().T
+    eigvals, v = cirq.unitary_eig(u1u2)
+    d = np.diag(np.sqrt(eigvals))
+
+    w = d @ v.conj().T @ u2
+
+    circuit_u1u2_mid = _middle_multiplexor_to_ops(q0, q1, q2, eigvals)
+
+    v = diagonal @ v
+
+    d_v, circuit_u1u2_r = opt.two_qubit_matrix_to_diagonal_and_operations(q1, q2, v, atol=atol)
+
+    w = d_v @ w
+
+    # if it's interesting to extract the diagonal then let's do it
+    if shift_left:
+        d_w, circuit_u1u2_l = opt.two_qubit_matrix_to_diagonal_and_operations(q1, q2, w, atol=atol)
+    # if we are at the end of the circuit, then just fall back to KAK
+    else:
+        d_w = np.eye(4)
+        circuit_u1u2_l = opt.two_qubit_matrix_to_operations(
+            q1, q2, w, allow_partial_czs=False, atol=atol
+        )
```

Comment:
This is a little awkward. For example, when we pass `shift=False` we ignore `d_w` so setting it to `np.eye(4)` is unnecessary (but the function signature forces us to return something...). It seems to me we can make it more elegant by moving the `if shift_left` conditional outside of this function. This also reduces the the number of arguments since it allows us to get rid of `shift_left`.

Reply:
I know it's a bit awkward but theoretically correct and keeps the conceptual boundary nice and clean: this method returns the operations for a two-qubit multiplexor and extracts a diagonal optionally. That optionality is expressed via the diagonal being identity or not. 

If I pull out the if statement - then we have a method that "extracts some operations from the two-qubit multiplexor, and returns the W matrix for further processing"... It will mix the levels of abstraction in my opinion. 

I think I'll have to push back on this one as well, unless you see a resolution that is clean conceptually as well.

Reply:
Sounds good - I agree about keeping clean mapping between code and concepts in the paper.

It still bothers me that we instantiate the identity matrix that the caller just ignores. Instantiation of this matrix creates an expectation in the mind of someone reading the code that it'll be used for something.

Suggestion: Let's use `None` in place of `np.eye(4)`. This has a nice ripple effect of changing the signature of the function by introducing explicit `Optional` that makes the optional character of the array explicit.

Reply:
I like that suggestion, thank you!